File: packages/sdk/php/src/ScopeManager.php
<?php

declare(strict_types=1);

namespace Bugwatch;

use Bugwatch\Types\Breadcrumb;
use Bugwatch\Types\Level;
use Bugwatch\Types\UserContext;

final class ScopeManager
{
    private static ?self $instance = null;

    private ?UserContext $user = null;

    /** @var array<string, string> */
    private array $tags = [];

    /** @var array<string, mixed> */
    private array $extras = [];

    /** @var Breadcrumb[] */
    private array $breadcrumbs = [];

    private int $maxBreadcrumbs = 100;

    private function __construct() {}

    public static function getInstance(): self
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }

        return self::$instance;
    }

    public static function reset(): void
    {
        self::$instance = null;
    }

    public function setUser(?UserContext $user): void
    {
        $this->user = $user;
    }

    public function getUser(): ?UserContext
    {
        return $this->user;
    }

    /**
     * @param array<string, string> $tags
     */
    public function setTags(array $tags): void
    {
        $this->tags = array_merge($this->tags, $tags);
    }

    /**
     * @return array<string, string>
     */
    public function getTags(): array
    {
        return $this->tags;
    }

    /**
     * @param array<string, mixed> $extras
     */
    public function setExtras(array $extras): void
    {
        $this->extras = array_merge($this->extras, $extras);
    }

    /**
     * @return array<string, mixed>
     */
    public function getExtras(): array
    {
        return $this->extras;
    }

    public function addBreadcrumb(Breadcrumb $breadcrumb): void
    {
        $this->breadcrumbs[] = $breadcrumb;

        if (count($this->breadcrumbs) > $this->maxBreadcrumbs) {
            $this->breadcrumbs = array_slice($this->breadcrumbs, -$this->maxBreadcrumbs);
        }
    }

    /**
     * @param array<string, mixed> $data
     */
    public function addBreadcrumbSimple(
        string $category,
        string $message = '',
        Level $level = Level::Info,
        array $data = [],
    ): void {
        $this->addBreadcrumb(Breadcrumb::create(
            category: $category,
            message: $message,
            level: $level,
            data: $data,
        ));
    }

    /**
     * @return Breadcrumb[]
     */
    public function getBreadcrumbs(): array
    {
        return $this->breadcrumbs;
    }

    public function clearBreadcrumbs(): void
    {
        $this->breadcrumbs = [];
    }

    public function setMaxBreadcrumbs(int $max): void
    {
        $this->maxBreadcrumbs = $max;
    }
}

File: packages/sdk/java/core/src/main/java/dev/bugwatch/ScopeManager.java
package dev.bugwatch;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Thread-local scope manager for maintaining contextual data (user, tags, extras, breadcrumbs)
 * across the call stack. Supports push/pop semantics for nested scopes.
 */
public class ScopeManager {

    private final ThreadLocal<Deque<Scope>> scopeStack = ThreadLocal.withInitial(() -> {
        Deque<Scope> stack = new ArrayDeque<>();
        stack.push(new Scope());
        return stack;
    });

    /**
     * Returns the current (top-of-stack) scope.
     */
    public Scope currentScope() {
        return scopeStack.get().peek();
    }

    /**
     * Pushes a clone of the current scope onto the stack.
     */
    public void pushScope() {
        Scope current = currentScope();
        scopeStack.get().push(current.copy());
    }

    /**
     * Pops the top scope. If only one scope remains, it is not popped.
     */
    public void popScope() {
        Deque<Scope> stack = scopeStack.get();
        if (stack.size() > 1) {
            stack.pop();
        }
    }

    public void setUser(UserContext user) {
        currentScope().user = user;
    }

    public void setTags(Map<String, String> tags) {
        currentScope().tags.putAll(tags);
    }

    public void setExtras(Map<String, Object> extras) {
        currentScope().extras.putAll(extras);
    }

    public void addBreadcrumb(Breadcrumb breadcrumb) {
        currentScope().breadcrumbs.add(breadcrumb);
    }

    /**
     * Applies the current scope data to an error event.
     */
    public void applyTo(ErrorEvent event) {
        Scope scope = currentScope();
        if (scope.user != null && event.getUser() == null) {
            event.setUser(scope.user);
        }
        if (!scope.tags.isEmpty()) {
            Map<String, String> merged = new HashMap<>(scope.tags);
            if (event.getTags() != null) {
                merged.putAll(event.getTags());
            }
            event.setTags(merged);
        }
        if (!scope.extras.isEmpty()) {
            Map<String, Object> merged = new HashMap<>(scope.extras);
            if (event.getExtra() != null) {
                merged.putAll(event.getExtra());
            }
            event.setExtra(merged);
        }
        if (!scope.breadcrumbs.isEmpty()) {
            List<Breadcrumb> merged = new ArrayList<>(scope.breadcrumbs);
            if (event.getBreadcrumbs() != null) {
                merged.addAll(event.getBreadcrumbs());
            }
            event.setBreadcrumbs(merged);
        }
    }

    /**
     * Internal scope data holder.
     */
    static class Scope {
        UserContext user;
        final Map<String, String> tags = new HashMap<>();
        final Map<String, Object> extras = new HashMap<>();
        final List<Breadcrumb> breadcrumbs = new ArrayList<>();

        Scope copy() {
            Scope copy = new Scope();
            copy.user = this.user;
            copy.tags.putAll(this.tags);
            copy.extras.putAll(this.extras);
            copy.breadcrumbs.addAll(this.breadcrumbs);
            return copy;
        }
    }
}
